Functional.Shared/DisposableBase.cs

```diff
@@ -0,0 +1,59 @@
+﻿using System;
+
+namespace Functional
+{
+	internal interface IDisposableObservable : IDisposable
+	{
+		/// <summary>
+		/// Gets a value indicating whether this object has been disposed.
+		/// </summary>
+		bool IsDisposed { get; }
+	}
+
+	/// <summary>
+	/// Implements the IDisposable pattern
+	/// </summary>
+	internal abstract class DisposableBase : IDisposableObservable
+	{
+		public bool IsDisposed { get; private set; }
+
+		/// <summary>
+		/// Finalizes an instance of the <see cref="DisposableBase"/> class.
+		/// </summary>
+		~DisposableBase()
+		{
+			Dispose(false);
+		}
+
+		/// <inheritdoc/>
+		public void Dispose()
+		{
+			Dispose(true);
+			GC.SuppressFinalize(this);
```

Comment:
This does nothing on classes that do not implement a finalizer, and I believe all public types in this library are either structs or static, so they can't be extended.

Reply:
This just follows IDisposable pattern in official doc:  https://docs.microsoft.com/en-us/dotnet/api/system.idisposable?view=netframework-4.7.2

Reply:
After giving this more thought, I think you are right that this pattern is needed. The finalizer is needed to ensure that any enumerators provided externally from the library are disposed correctly. That being said, I'm still unsure of what to do about the async enumerators. Maybe dispose them in a continuation?

Reply:
@JohannesMoersch without async await, maybe only in continuation : )